File: 5-longest-palindromic-substring.cpp
#include <bits/stdc++.h>
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        int dp_len[n][n];
        for(int r = 0; r < n; ++r)
            for(int c = 0; c < n; ++c)
                dp_len[r][c] = 0;
        for(int r = 0; r < n; ++r)
            dp_len[r][r] = 1;

        int max_len = 1;
        int max_r = 0;
        int max_c = 0;
        for (int l = 1; l < n; ++l){
            for (int r = 0; r < n-l; ++r){
                int c = r + l;
                if (s[r] == s[c] && ((c - r < 2) || dp_len[r + 1][c - 1] > 0)){
                    dp_len[r][c] = 2 + dp_len[r + 1][c - 1];
                    if (dp_len[r][c] > max_len){
                        max_len = dp_len[r][c];
                        max_r = r;
                        max_c = c;
                    }
                }
            }
        }
        return s.substr(max_r, max_len);
    }
};

File: 5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        dp = [[False]*(len(s)+1) for _ in range(len(s)+1)]
        
        max_len = 1
        max_start = 0
        for i in range(1, len(s)+1):
            dp[i][i] = True
        for l in range(2, len(s)+1):
            for i in range(1, len(s)-l+1+1):
                j = i + l - 1
                if s[i-1]==s[j-1]:
                    dp[i][j] = (l<3) or dp[i+1][j-1]
                    if dp[i][j] and l>max_len:
                        max_len = l
                        max_start = i-1
                # print(i,j, s[i-1], s[j-1], dp[i][j], [l_i])
                # print()
        # for i in range(len(s)+1):
        #     for j in range(len(s)+1):
        #         print(dp[i][j], end=' ')
        #     print()

        return s[max_start:max_start+max_len]
